File: codingTest/kotlin/algorithm/src/three/수 찾기.kt
package three

fun main() {
    readLine()!!
    val A = readLine()!!.split(' ').map {it.toInt()}.toIntArray()
    readLine()!!
    val find = readLine()!!.split(' ').map {it.toInt()}
    A.sort()
    for (el in find) binarySearch(el, A)
}

fun binarySearch(num: Int, arr: IntArray) {
    var low = 0
    var high = arr.size-1
    while (low <= high) {
        var middle = (low + high) / 2
        if (num == arr[middle]) return println(1)
        else if (num < arr[middle]) {
            high = middle - 1
            middle = (low+high)/2
        }
        else {
            low = middle + 1
            middle = (low+high)/2
        }
    }
    return println(0)
}


//fun search(el: Int, list: List<Int>): Int {
//    var listSet = list.toSet()
//    listSet.find { it == el }?: return 0
//    return 1
//}


//fun search(el: Int, A: List<Int>): Int {
//    val list = A.sorted()
//    val middle = A.size / 2
//    return if (list.size == 1 && list[0] != el) 0
//    else if (el == list[middle]) 1
//    else if (el < list[middle]) search(el, list.subList(0, middle))
//    else search(el, list.subList(middle, list.size))
//}

File: codingTest/javascript/three/수 찾기.js
const rl = require("readline").createInterface({ input: process.stdin, output: process.stdout });
var list = [];
rl.on("line", (input) => {
  list.push(input);
}).on("close", () => {
  findNumber(list);
});

function findNumber(list) {
  var N = Number(list[0]);
  var arr = list[1].split(" ").map((e) => Number(e));
  var M = Number(list[2]);
  var searchList = list[3].split(" ").map((e) => Number(e));
  arr.sort((a, b) => a - b);
  for (let num of searchList) console.log(binarySearch(num, arr));
}

function binarySearch(num, arr) {
  let low = 0;
  let high = arr.length - 1;
  while (low <= high) {
    let mid = Math.floor((high + low) / 2);
    if (arr[mid] === num) return 1;
    else if (arr[mid] > num) high = mid - 1;
    else low = mid + 1;
  }
  return 0;
}
